### PR TITLE
Add multi-search federation

### LINKED ISSUE #573: [v1.10.0] Federated search
*Following this [central issue](https://github.com/meilisearch/integration-guides/issues/302)*

### Explanation of the feature

Pass new search parameters to `multiSearch` method to get one single list of result
Ensure the queries passed can accept the `federationOptions` search parameter
No need to handle the case of errors, Meilisearch does it

Usage: https://meilisearch.notion.site/v1-10-federated-search-698dfe36ab6b4668b044f735fb40f0b2?pvs=74

### TODO

- [ ] Add a new search parameter named `federation` to the `multiSearch` method
- [ ] Add the capability to have a new search parameter to the search query named `federationOptions`
- [ ] Add tests

⚠️ Please do 1 PR for all of these changes, and not several.

### PR DESCRIPTION
**Pull request by polyfloyd**
# Pull Request

Fixes #573

## What does this PR do?
This PR adds support for the new federated search to be introduced in v1.10: https://github.com/meilisearch/meilisearch/releases/tag/v1.10.0-rc.0

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

### CODE REVIEW

**Comment by polyfloyd** [comment]
Please note that I am running into an issue here: `unaccepted status code found: 400 expected: [200], MeilisearchApiError Message: Inside .queries[0]: Using pagination options is not allowed in federated queries.`

I think this patch is correct, so I'm opening this PR already. But I can not be 100% sure just yet

**Comment by curquiza** [comment]
Hello @polyfloyd 
thanks for your PR, but v1.10 is not released yet. So I will review and merge a PR for federated once v1.10 is release 😊 

Thanks for your anticipation and involvement 😄

**Comment by polyfloyd** [comment]
Found the cause of the error message, this Go library sets the pagination limit of each query to a default. I removed it for now so I can continue testing

**Comment by Ja7ad** [comment]
@polyfloyd Thank you, currently you can change this PR to draft until release v1.10 and add issue for federation search.

**Comment by polyfloyd** [comment]
Saw that there is an issue open now: #573

Will go over the checklist to see whether there is still something to be done

**Comment by Ja7ad** [comment]
> Saw that there is an issue open now: #573
> 
> Will go over the checklist to see whether there is still something to be done

Please do this check list:

- [ ] write test stage for multi-search with federation
- [ ] add `_federation` for `SearchResponse`

```json
"_federation": {
        "indexUid": "comics",
        "queriesPosition": 1
      }
```

**Comment by polyfloyd** [comment]
Thanks!

File: types.go [hunk 1]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
 
 type MultiSearchResponse struct {
 	Results []SearchResponse `json:"results"`
+	// For federated results.
```
Comment (Ja7ad) [inline types.go:448]: Please remove comment

File: types.go [hunk 2]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
 
 type MultiSearchResponse struct {
 	Results []SearchResponse `json:"results"`
```
Comment (Ja7ad) [inline types.go:447]: Please add tag `omitempty`

File: types.go [hunk 3]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
...
+	// For federated results.
+	Hits               []interface{} `json:"hits"`
```
Comment (Ja7ad) [inline types.go:449]: Please add tag `omitempty`

File: types.go [hunk 4]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
...
+	ProcessingTimeMs   int64         `json:"processingTimeMs"`
```
Comment (Ja7ad) [inline types.go:450]: Please add tag `omitempty`

File: types.go [hunk 5]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
...
+	Offset             int64         `json:"offset"`
```
Comment (Ja7ad) [inline types.go:451]: Please add tag `omitempty`

File: types.go [hunk 6]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
...
+	Limit              int64         `json:"limit"`
```
Comment (Ja7ad) [inline types.go:452]: Please add tag `omitempty`

File: types.go [hunk 7]
```diff
@@ -434,6 +445,13 @@ type SearchResponse struct {
...
+	EstimatedTotalHits int64         `json:"estimatedTotalHits"`
+	SemanticHitCount   int64         `json:"semanticHitCount"`
```
Comment (Ja7ad) [inline types.go:454]: Please add tag `omitempty`

**Review by Ja7ad — CHANGES_REQUESTED** [review]

**Comment by Ja7ad** [comment]
> Updated the commit title and added a test.
> 
> Adding `_federation` to SearchResponse does not seem correct? This object is present in each item of the `hits` array. See the expected payload of the new unit test to see what I mean

```go
type SearchResponse struct {
	Hits               []interface{}       `json:"hits"`
	EstimatedTotalHits int64               `json:"estimatedTotalHits,omitempty"`
	Offset             int64               `json:"offset,omitempty"`
	Limit              int64               `json:"limit,omitempty"`
	ProcessingTimeMs   int64               `json:"processingTimeMs"`
	Query              string              `json:"query"`
	FacetDistribution  interface{}         `json:"facetDistribution,omitempty"`
	TotalHits          int64               `json:"totalHits,omitempty"`
	HitsPerPage        int64               `json:"hitsPerPage,omitempty"`
	Page               int64               `json:"page,omitempty"`
	TotalPages         int64               `json:"totalPages,omitempty"`
	FacetStats         interface{}         `json:"facetStats,omitempty"`
	IndexUID           string              `json:"indexUid,omitempty"`
	Federation         *FederationResponse `json:"_federation,omitempty"`
}

type FederationResponse struct {
	IndexUID        string `json:"indexUid,omitempty"`
	QueriesPosition int64  `json:"queriesPosition,omitempty"`
}
```

**Comment by Ja7ad** [comment]
> Updated the commit title and added a test.
> 
> Adding `_federation` to SearchResponse does not seem correct? This object is present in each item of the `hits` array. See the expected payload of the new unit test to see what I mean

Don't require any do for this.

**Review by Ja7ad — APPROVED** [review]

**Comment by Ja7ad** [comment]
@polyfloyd please add `fixes #573` in PR description

**Comment by polyfloyd** [comment]
@Ja7ad I have added the issue ref to the PR description

**Comment by Ja7ad** [comment]
bors merge

Thank you!!

**Pull request merged**